File: frontend_flutter/lib/services/crypto_service.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import 'package:cryptography/cryptography.dart';

/// Client-side encryption service implementing zero-knowledge architecture.
/// Uses Argon2id for key derivation and AES-256-GCM for encryption.
class CryptoService {
  static final CryptoService _instance = CryptoService._internal();
  factory CryptoService() => _instance;
  CryptoService._internal();

  Uint8List? _masterKey;
  final _random = Random.secure();

  /// Derive a 256-bit master key from the user's master password using PBKDF2.
  Future<Uint8List> deriveKey(String masterPassword, String email) async {
    final algorithm = Pbkdf2(
      macAlgorithm: Hmac.sha256(),
      iterations: 100000,
      bits: 256,
    );

    // Use email as salt for deterministic key derivation
    final salt = utf8.encode(email.toLowerCase().trim());

    final result = await algorithm.deriveKey(
      secretKey: SecretKey(utf8.encode(masterPassword)),
      nonce: salt,
    );

    final keyBytes = await result.extractBytes();
    _masterKey = Uint8List.fromList(keyBytes);
    return _masterKey!;
  }

  /// Encrypt plaintext JSON data using AES-256-GCM.
  /// Returns Base64-encoded string: iv(12) + ciphertext + mac(16)
  Future<String> encrypt(Map<String, dynamic> data) async {
    if (_masterKey == null) throw StateError('Master key not derived');

    try {
      final plaintext = utf8.encode(jsonEncode(data));
      final algorithm = AesGcm.with256bits();
      final secretKey = SecretKey(_masterKey!);
      final nonce = algorithm.newNonce(); // 12 bytes IV

      final secretBox = await algorithm.encrypt(
        plaintext,
        secretKey: secretKey,
        nonce: nonce,
      );

      // WebCrypto standard: IV + CIPHERTEXT + MAC
      final combined = <int>[
        ...secretBox.nonce,
        ...secretBox.cipherText,
        ...secretBox.mac.bytes,
      ];

      return base64Encode(combined);
    } catch (e) {
      print('[CryptoService] Encryption error: $e');
      rethrow;
    }
  }

  /// Decrypt Base64-encoded encrypted data back to JSON.
  Future<Map<String, dynamic>> decrypt(String encryptedBase64) async {
    if (_masterKey == null) throw StateError('Master key not derived');

    try {
      final combined = base64Decode(encryptedBase64);

      // Extract: iv(12) + ciphertext + mac(16)
      // WebCrypto combined block = IV (12) + CT (var) + MAC (16)
      final nonce = combined.sublist(0, 12);
      final macBytes = combined.sublist(combined.length - 16);
      final cipherText = combined.sublist(12, combined.length - 16);

      final secretBox = SecretBox(
        cipherText,
        nonce: nonce,
        mac: Mac(macBytes),
      );

      final algorithm = AesGcm.with256bits();
      final secretKey = SecretKey(_masterKey!);

      final plaintext = await algorithm.decrypt(secretBox, secretKey: secretKey);
      final jsonString = utf8.decode(plaintext);

      return jsonDecode(jsonString) as Map<String, dynamic>;
    } catch (e) {
      // Fallback to legacy Base64 for development migration
      try {
        final decodedStr = utf8.decode(base64Decode(encryptedBase64));
        return jsonDecode(decodedStr) as Map<String, dynamic>;
      } catch (legacyErr) {
        print('[CryptoService] Decryption error (incl fallback): $e');
        rethrow;
      }
    }
  }

  /// Generate a random password.
  String generatePassword({
    int length = 20,
    bool uppercase = true,
    bool lowercase = true,
    bool numbers = true,
    bool symbols = true,
  }) {
    const upperChars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    const lowerChars = 'abcdefghijklmnopqrstuvwxyz';
    const numChars = '0123456789';
    const symbolChars = '!@#\$%^&*()_+-=[]{}|;:,.<>?';

    String chars = '';
    if (uppercase) chars += upperChars;
    if (lowercase) chars += lowerChars;
    if (numbers) chars += numChars;
    if (symbols) chars += symbolChars;

    if (chars.isEmpty) chars = lowerChars + numChars;

    return List.generate(length, (_) => chars[_random.nextInt(chars.length)]).join();
  }

  bool get hasKey => _masterKey != null;

  String? get masterKeyBase64 {
    return _masterKey != null ? base64Encode(_masterKey!) : null;
  }

  void setMasterKeyBase64(String base64Key) {
    _masterKey = Uint8List.fromList(base64Decode(base64Key));
  }

  void clearKey() {
    if (_masterKey != null) {
      _masterKey!.fillRange(0, _masterKey!.length, 0);
      _masterKey = null;
    }
  }
}

File: frontend_web/src/services/CryptoService.js
/**
 * CryptoService — Client-side encryption for SecureVault
 * Standardized across Web, Extension, and Mobile.
 * Uses AES-256-GCM for data and PBKDF2-HMAC-SHA256 for key derivation.
 */

class CryptoService {
  constructor() {
    this.masterKey = null; // Stored in memory after login
  }

  /**
   * Derives a 256-bit master key from password and email (as salt).
   */
  async deriveMasterKey(password, email) {
    const encoder = new TextEncoder();
    const passwordBytes = encoder.encode(password);
    const saltBytes = encoder.encode(email.toLowerCase().trim());

    // Import the password as a raw key
    const baseKey = await window.crypto.subtle.importKey(
      'raw',
      passwordBytes,
      { name: 'PBKDF2' },
      false,
      ['deriveKey']
    );

    // Derive the master key
    this.masterKey = await window.crypto.subtle.deriveKey(
      {
        name: 'PBKDF2',
        salt: saltBytes,
        iterations: 100000,
        hash: 'SHA-256'
      },
      baseKey,
      { name: 'AES-GCM', length: 256 },
      true, // Must be true to extract/export if needed, but false is safer if only used internally
      ['encrypt', 'decrypt']
    );

    return this.masterKey;
  }

  /**
   * Encrypts a JSON object.
   * Returns: Base64(iv(12) + ciphertext + tag(16))
   */
  async encrypt(data) {
    if (!this.masterKey) throw new Error('Vault is locked (no master key)');

    const encoder = new TextEncoder();
    const plaintext = encoder.encode(JSON.stringify(data));
    const iv = window.crypto.getRandomValues(new Uint8Array(12));

    const encryptedContent = await window.crypto.subtle.encrypt(
      { name: 'AES-GCM', iv },
      this.masterKey,
      plaintext
    );

    const combined = new Uint8Array(iv.length + encryptedContent.byteLength);
    combined.set(iv, 0);
    combined.set(new Uint8Array(encryptedContent), iv.length);

    return btoa(String.fromCharCode(...combined));
  }

  /**
   * Decrypts a Base64-encoded combined block.
   */
  async decrypt(encryptedBase64) {
    if (!this.masterKey) throw new Error('Vault is locked (no master key)');

    const combined = new Uint8Array(
      atob(encryptedBase64)
        .split('')
        .map(c => c.charCodeAt(0))
    );

    const iv = combined.slice(0, 12);
    const ciphertext = combined.slice(12);

    const decryptedContent = await window.crypto.subtle.decrypt(
      { name: 'AES-GCM', iv },
      this.masterKey,
      ciphertext
    );

    const decoder = new TextDecoder();
    return JSON.parse(decoder.decode(decryptedContent));
  }

  setMasterKey(key) {
    this.masterKey = key;
  }

  clearKey() {
    this.masterKey = null;
  }
}

export default new CryptoService();
